chore(cli): re-order CLI create command

DanielleMaywood opened the pull request:
Relates to https://github.com/coder/internal/issues/893

Instead of `coder task create <template> --input <input>`, it is now `coder task create <input> --template <template>`.

If there is only one AI task template on the deployment, the `--template` parameter can be omitted.

DanielleMaywood committed:
chore(cli): re-order CLI create command

Relates to https://github.com/coder/internal/issues/893

Instead of `coder task create <template> --input <input>`, it is now
`coder task create <input> --template <template>`.

If there is only one AI task template on the deployment, the `--template`
parameter can be omitted.

mafredri reviewed (commented):
I think this new approach will ultimately be better, thanks for implementing! Some feedback inline.

matifali reviewed (approved):
Thanks <3. 🚢 

Should we mark this as breaking change?

johnstcn reviewed (approved):
Nothing to add beyond what @mafredri already mentioned. Nice work! 👍

matifali commented on cli/exp_task_create.go:
```diff
@@ -24,24 +23,25 @@ func (r *RootCmd) taskCreate() *serpent.Command {
-				Flag:     "input",
-				Env:      "CODER_TASK_INPUT",
-				Value:    serpent.StringOf(&taskInput),
-				Required: true,
```
Maybe, yes. Prebuilds are tricky.

mafredri replied:
I don't think we should start tasks without input necessarily, but who's to say there isn't a default input that's only controlled by e.g. template parameters. This was one motivation that `--input` was made a flag previously.

matifali commented:
> > Should we mark this as breaking change?
> 
> No need, this is exactly why we placed these under `exp`. No announcement for release yet either.

But many of our customers are using Tasks. So I was just being extra cautious not to break their workflows without any information.

DanielleMaywood commented on cli/exp_task_create.go:
```diff
@@ -67,20 +67,28 @@ func (r *RootCmd) taskCreate() *serpent.Command {
+				// allow omitting the template. Otherwise we will require
+				// the user to be explicit with their choice of template.
+				if len(templates) > 1 {
+					return xerrors.Errorf("template name not provided")
```
How should we decide what templates go here? If there are a lot of task templates this could look quite gnarly

mafredri commented on cli/exp_task_create.go:
```diff
@@ -24,24 +23,25 @@ func (r *RootCmd) taskCreate() *serpent.Command {
-				Required: true,
+				Flag:  "input",
+				Env:   "CODER_TASK_INPUT",
+				Value: serpent.StringOf(&taskInput),
```
If we read from stdin we should also allow zero args. I don't think we have much precedent but I'm not opposed as it's a more natural way to input multi line. 👍🏻

`cat make-it-so.txt | coder task create`.

DanielleMaywood replied:
Should we copy how the `org create` command works?

It accepts a flag `--stdin` to indicate it will read from stdin

https://github.com/coder/coder/blob/3071ef8a19670fc5a69733cf5107283f14a00df2/cli/organizationroles.go#L138-L144

mafredri replied:
Makes sense to me. This command (could be) interactive (prompt yes/no) so explicitly requiring stdin flag would also enforce `--yes` (can be documented).

DanielleMaywood commented:
> (@DanielleMaywood, did create slip into the release?)

It was merged on Friday so I don't think so

mafredri commented:
> It was merged on Friday so I don't think so

Thanks, then I don't see any reason to tag this breaking. 👍🏻

mafredri commented on cli/exp_task_create.go:
```diff
@@ -90,6 +98,15 @@ func (r *RootCmd) taskCreate() *serpent.Command {
 			}
 
+			if templateVersionName != "" {
+				templateVersion, err := client.TemplateVersionByOrganizationAndName(ctx, organization.ID, templateName, templateVersionName)
```
`templateVersionID = templates[0].ActiveVersionID` 😄

DanielleMaywood committed:
refactor: remove possibility of unneeded database call.

In the scenario where both a template name and a template version were
provided, we'd hit the database twice, depsite only needing to do it
once.

DanielleMaywood committed:
feat: add `--stdin` flag to allow reading from stdin for task input

DanielleMaywood committed:
chore: disallow empty task input

DanielleMaywood committed:
test: fix tests

DanielleMaywood committed:
chore: appease the linter

DanielleMaywood committed:
chore: add detail to error message

mafredri reviewed (approved):
Nice work!
